app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java

```diff
@@ -1465,6 +1466,36 @@ public void signUpNewUser_emptyPassword_signUpFails() {
         assertTextOnScreen(PASSWORD_IS_REQUIRED);
     }
 
+    @Test
+    public void signUpNewUser_usernameExists_emailExists_signUpFails() {
+        User newUser = new User(TEST_USER_ID, TEST_USER_USERNAME, TEST_USER_EMAIL, VALID_PASSWORD);
+
+        _solo.clickOnButton(SIGN_UP_TEXT);
+
+        _solo.enterText(_solo.getEditText("Username"), newUser.getUsername());
+        _solo.enterText(_solo.getEditText("Email"), newUser.getEmail());
+        _solo.enterText(_solo.getEditText("Password"), newUser.getPassword());
+
+        _solo.clickOnButton(SIGN_UP_TEXT);
+
+        assertTextOnScreen(THIS_USERNAME_IS_ALREADY_TAKEN_MESSAGE);
+    }
+
+    @Test
+    public void signUpNewUser_usernameExists_newEmail_signUpFails() {
+        User newUser = new User(TEST_USER_ID, TEST_USER_USERNAME, "newemail@gmail.com", VALID_PASSWORD);
+
+        _solo.clickOnButton(SIGN_UP_TEXT);
+
+        _solo.enterText(_solo.getEditText("Username"), newUser.getUsername());
+        _solo.enterText(_solo.getEditText("Email"), newUser.getEmail());
+        _solo.enterText(_solo.getEditText("Password"), newUser.getPassword());
+
+        _solo.clickOnButton(SIGN_UP_TEXT);
+
+        assertTextOnScreen(THIS_USERNAME_IS_ALREADY_TAKEN_MESSAGE);
+    }
```

Comment:
2 new UI tests:
* Sign up with a taken username and taken email, sign up fails
* Sign up with a taken username and new email, sign up fails